### ISSUE TITLE
Support for GPT-4 Turbo with 128k context

### ISSUE THREAD

**Issue by AW2307** [brianpetro/obsidian-smart-connections#372]
Hi,

Since a preview version of the new GPT4 Turbo model (gpt-4-1106-preview)with 128k context is already accessible via the API, could you please enable using this model in SmartConnections?

Cheers.

**Comment by quelsemme** [comment]
seconded!

**Comment by AntonSoeffing** [comment]
Yes please! Can someone capable create a pull request for it maybe?

**Comment by mashean** [comment]
Yes...it would make an already awesome tool even better!

**Comment by brianpetro** [comment]
@AntonSoeffing @mashean @AW2307 @quelsemme The latest update `1.6.37` includes GPT-4 turbo. I also made a few tweaks to better leverage the large context window. Let me know if you run into any issues or if it works as expected.

🌴

**Comment by quelsemme** [comment]
Fantastic @brianpetro. Love your work. Will check it out.

**Comment by brianpetro** [comment]
Note: keep an eye on your usage costs when using `gpt-4-turbo`.

Between the higher context limit and higher cost per token, it can add up quickly!

🌴

**Comment by AW2307** [comment]
Many thanks for adding this!

Cost per token is actually lower than GPT4, but with the added context it probably still adds up quicker.

**Comment by brianpetro** [comment]
> Cost per token is actually lower than GPT4, but with the larger context it probably still adds up quicker.

@AW2307 yes, that's exactly it.

**Comment by clappo143** [comment]
@brianpetro Thanks very much for the update! Haven't used it much, but initial impressions are that GPT4 turbo results in notably improved performance/outputs :)  

Re the above, it seems that with GPT4 turbo selected, Smartconnections will retrieve 30 notes (or the embeddings, I guess) for the context it uses to inform its response. When I copy and paste it, I note that there is a fair bit of duplication (red highlights below): 

![image](https://github.com/brianpetro/obsidian-smart-connections/assets/95749676/572b63af-1d5e-4096-a0d2-b02ec553d9a4)

The query was about something for which I have perhaps 10-20 relevant notes in my vault. I'm wondering if it is retrieving identical embeddings/notes to reach the '30' quota? Whatever the case, it seems not great in terms of token usage (though thinking about it, perhaps it actually is useful, in that it reinforces the context - or at least certain parts of it..)?

Anyway, just an observation that seemed potentially relevant to the comments about the larger context and token usage. 

Thanks again for the update (and building the plugin too)!

**Comment by brianpetro** [comment]
@clappo143 thanks for sharing that!

I can't see it in the screenshot, but the context items should be labeled by which note they came from on the lines that start `---BEGIN`. Reviewing that might help clarify what's happening.

One possibility is that both sections of a note and the entire note are included.

While I'm reserving any future v1 development for major issues, I do plan on addressing this in v2 and just added this specific improvement (preventing duplicate context) to my shortlist of context retrieval updates.

Thanks for bringing this to my attention. This sort of feedback goes a long way in helping development!

🌴

**Comment by clappo143** [comment]
@brianpetro No problem at all! 

After reading your response and doing a couple more tests (and actually reviewing the pasted context this time), it seems that it is in fact retrieving unique notes for context – no duplicates. Turns out that a lot of my notes are different versions/drafts of the same document, or poorly organised research notes comprised in part of identical batches of copy-and-pasted material – so while it is actually retrieving unique notes for the context, some of those notes happen to contain some identical content as others. I hope that makes sense?! 

From my perspective, removing any identical material from the context would be nice – as I think models like GPT4 are sophisticated enough to just work with the context, even if some notes have been manipulated/purged to avoid duplicate content. But, I think having the amount of duplicate material in the context as in that screenshot I shared might be more unique to me and the nature of the notes in my vault, rather than reflecting an actual problem that needs to be fixed! 

Also, from the few tests I did, I'm quite confident that it is not trying to retrieve 30 notes regardless of whether the vault has that many notes that are actually relevant to the query. For obscure questions, just 2-3 notes were retrieved as context 👍

Hope that helps! :)
